src/main/java/com/iota/iri/Snapshot.java

```diff
@@ -2,175 +2,181 @@
+
+            Map<Hash, Long> state = new HashMap<>();
+
+            reader.lines().forEach(line -> {
+                int index = line.indexOf(';');
+                if (index != -1 && index != line.length() - 1) {
+                    String key = line.substring(0, index);
+                    String value = line.substring(index + 1);
                     state.put(new Hash(key), Long.valueOf(value));
                 }
-            }
-        } catch (IOException e) {
-            //syso is left until logback is fixed
-            System.out.println("Failed to load snapshot.");
-            log.error("Failed to load snapshot.", e);
-            System.exit(-1);
+            });
+            return state;
         }
-        finally {
-            IOUtils.closeQuietly(reader);
-        }
-        return state;
     }
 
-    protected final Map<Hash, Long> state;
-    private int index;
+    private Snapshot(Map<Hash, Long> state, int index, boolean immutable) {
+        this.immutable = immutable;
+        this.state = immutable ? Collections.unmodifiableMap(new HashMap<>(state)) : new HashMap<>(state);
+        this.index = index;
+    }
 
-    public int index() {
-        int i;
-        rwlock.readLock().lock();
-        i = index;
-        rwlock.readLock().unlock();
-        return i;
+    Map<Hash, Long> copyState() { // VISIBLE FOR TESTING ONLY
+        return withReadLockSupply(() -> new HashMap<>(state));
     }
 
-    private Snapshot(Map<Hash, Long> initialState, int index) {
-        state = new HashMap<>(initialState);
-        this.index = index;
+    public int index() {
+        return withReadLockSupply(() -> index);
     }
 
-    public Snapshot clone() {
-        return new Snapshot(state, index);
+    /**
+     * @return a modifiable copy of this.
+     */
+    public Snapshot copySnapshot() {
+        return withReadLockSupply(() -> new Snapshot(state, index, false));
     }
 
     public Long getBalance(Hash hash) {
-        Long l;
-        rwlock.readLock().lock();
-        l = state.get(hash);
-        rwlock.readLock().unlock();
-        return l;
+        return withReadLockSupply(() -> state.get(hash));
     }
 
-    public Map<Hash, Long> patchedDiff(Map<Hash, Long> diff) {
-        Map<Hash, Long> patch;
-        rwlock.readLock().lock();
-        patch = diff.entrySet().stream().map(hashLongEntry ->
-            new HashMap.SimpleEntry<>(hashLongEntry.getKey(), state.getOrDefault(hashLongEntry.getKey(), 0L) + hashLongEntry.getValue())
-        ).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        rwlock.readLock().unlock();
-        return patch;
+    Map<Hash, Long> patchedDiff(Map<Hash, Long> diff) {
+        return withReadLockSupply(() -> {
+            Map<Hash, Long> map = new HashMap<>(diff.size());
+            diff.forEach((key, value) -> {
+                Long previous = state.getOrDefault(key, 0L);
+                map.put(key, Math.addExact(previous, value));
```

Comment:
:grinning: 
looks much better than before

Reply:
This is the first time I see u use an emoji 😀